File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/tenant/TenantId.java
package pe.joedayz.microservicios.modulo01.tenant;

import java.util.Objects;

/**
 * Identificador de un tenant (tienda) en la plataforma SaaS multi-tenant.
 *
 * <p>Es un Value Object: inmutable y se compara por valor. Todo dato de negocio
 * (pedidos, catalogo, stock) pertenece a un tenant y jamas se cruza entre tenants.
 */
public record TenantId(String value) {

    public TenantId {
        Objects.requireNonNull(value, "tenantId no puede ser null");
        if (value.isBlank()) {
            throw new IllegalArgumentException("tenantId no puede estar vacio");
        }
    }

    public static TenantId of(String value) {
        return new TenantId(value);
    }

    @Override
    public String toString() {
        return value;
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Tenant/TenantId.cs
namespace JoeDayz.Microservicios.Modulo01.Tenant;

/// <summary>
/// Identificador de tenant. Value object: sin identidad propia, se compara por valor.
/// En un e-commerce multi-tenant TODA operacion viaja acompaniada de un TenantId.
/// </summary>
public sealed record TenantId
{
    public TenantId(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value, nameof(value));
        Value = value;
    }

    public string Value { get; }

    public static TenantId Of(string value) => new(value);

    public override string ToString() => Value;
}
